### DISCUSSION TITLE
Difficulty setting Spectral Output Bands

### DISCUSSION THREAD

**Discussion by coult099** [mitsuba-renderer/mitsuba3#1146] [Q&A] [answered]
I am attempting to render a scene with the `llvm_ad_spectral` variant of Mitsuba. My scene is in `.xml` format, and I would like to generate a 30-band multispectral image with linearly-spaced color channels from 368 - 830 nm. I have attempted to do this by using 
```
        <film type="specfilm" >
            <spectrum name="band1" value="368"/>
            <spectrum name="band2" value="383.93"/>
            ...
            <spectrum name="band30" value="830"/>
        </film >
```

However, when I do this `scene = mi.load_file(<my-xml>)` I get the error: 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: ​[xml.cpp:1083] Error while loading "./scenes/<my-xml>" (near line 91, col 10): could not instantiate film plugin of type "specfilm": jit_var_literal(): tried to create an array with 18446744073709551615 entries, which exceeds the limit of 2^32 == 4294967296 entries.
```

I can render with the spectral variant and output to an RGB image using `hdrfilm`, but I would like to use `specfilm` to be able to output a 30-channel image. Can anyone provide insight as to why I'm getting this error? Thank you!

System:
- Apple M2 Pro 
- macOS Sonoma 14.4.1
- 16 GB unified memory
- LLVM@16
- Mitsuba variant: `llvm_ad_spectral`

**Comment by mcrescas** [comment, marked as answer]
Hi,

The current implementation of the `specfilm` does not support discrete wavelengths for bands (which is the error that you are seeing). 
For your specific use case, I recommend you modify the film to account for the discrete nature of your wavelengths (you will have to change the sampling mechanism to avoid the creation of a continuous distribution).

Good luck!

**Reply by coult099** [reply]
Thank you for the reply! I did not know that spectral film didn't support discrete wavelength channels for output.

If I wanted instead to output a standard 3-channel RGB image, should I use `specfilm` or `hdrfilm` for that? I would still like to use the `spectral` variant of Mitsuba and specify that the sensor spectrum response should be uniform across 368 - 830 nm. 

Also, if I wanted to output a black and white image with single-wavelength information, is that possible using `specfilm`?

**Reply by mcrescas** [reply]
If you only want to render RGB curves in spectral, I recommend you to use one of the spectral variants of Mitsuba with the `hdrfilm` (it is the default behavior). Note that in that case, we do importance sampling of the spectral domain, so you are not covering uniformly the visible domain but instead focusing more samples depending on the RGB spectral sensitivity.

Regarding the second question: for that, you will have to modify a bit the `specfilm` to "ignore" the importance sampling computation that combines information for the "tentative" multiple SRFs.

**Reply by coult099** [reply]
Thank you! 

Just so I am understanding: if I want to render with `spectral` and have the sensor measure uniformly across 368 - 830 nm, and output a resulting image with 30 spectral color channels, I will have to modify `specfilm` to account for that? I think it's fine that the sensor samples from a continuous distribution, but I just want to be able to output discrete "cuts" from the spectrum. 

Edit: I ask this because in a previous version of Mitsuba (0.5.0), it was possible to use `hdrfilm` or `mfilm` with `pixelFormat=spectrum` and compile the system with a set number of spectrum samples, to output `N` color channels across 360-830nm. I am wondering if there is any similar technique existing in Mitsuba3 without having to edit the standard classes. Thanks!
